src/backend/joanie/core/models/products.py

```diff
@@ -756,13 +789,25 @@ def clean(self):
                 ]
                 if len(validated_user_orders) == 0:
                     message = _(
-                        f'Course run "{self.course_run.resource_link:s}" '
+                        f'Course run "{str(self.course_run.id)}" '
                         "requires a valid order to enroll."
                     )
                     raise ValidationError({"__all__": [message]})
             else:
                 message = _("You are not allowed to enroll to a course run not listed.")
                 raise ValidationError({"__all__": [message]})
+        elif (
+            not self.course_run.course.targeted_by_products.exists()
+            and self.was_created_by_order is True
```

Comment:
It's not clear at first sight if the not will be for both or for the first :thinking: 
You should check the second clause first so that you avoid the db query if it is false...